src/inc/til/spsc.h

```diff
@@ -0,0 +1,530 @@
+            }
+
+            T* data() const noexcept
+            {
+                return _data;
+            }
+
+        private:
+            void drop(atomic_size_type& mine)
+            {
+                // Signal the other side we're dropped. See acquire() for the handling of the drop_flag.
+                // We don't need to use release ordering like release() does as each call to
+                // any of the sender/receiver methods already results in a call to release().
+                // Another release ordered write can't possibly synchronize any more data anyways at this point.
+                const auto myPos = mine.load(std::memory_order_relaxed);
+                mine.store(myPos | drop_flag, std::memory_order_relaxed);
+                mine.notify_one();
+
+                // The first time SPSCBase is dropped (destroyed) we'll set
+                // the flag to true and get false, causing us to return early.
+                // Only the second time we'll get true.
+                // --> The contents are only deleted when both sides have been dropped.
+                if (_eitherSideDropped.exchange(true))
+                {
+                    delete this;
+                }
+            }
+
+            // NOTE: waitMask MUST be either 0 (consumer) or revolution_flag (producer).
+            acquisition acquire(atomic_size_type& mine, atomic_size_type& theirs, size_type waitMask, size_type slots) noexcept
+            {
+                size_type myPos = mine.load(std::memory_order_relaxed);
+                size_type theirPos;
+
+                while (true)
+                {
+                    // This acquire read synchronizes with the release write in release().
+                    theirPos = theirs.load(std::memory_order_acquire);
+                    if ((myPos ^ theirPos) != waitMask)
+                    {
+                        break;
+                    }
+
+                    theirs.wait(theirPos, std::memory_order_relaxed);
+                }
+
+                // If the other side's position contains a drop flag, as a X -> we need to...
+                // * producer -> stop immediately
+                //   Only the producer's waitMask is != 0.
+                // * consumer -> finish consuming all values and then stop
+                //   We're finished if the only difference between our
+                //   and the other side's position is the drop flag.
+                if ((theirPos & drop_flag) != 0 && (waitMask != 0 || (myPos ^ theirPos) == drop_flag))
+                {
+                    // An empty acquire struct is equal to false,
+                    // which signals that the other side has been dropped.
+                    return {};
+                }
+
+                auto begin = myPos & position_mask;
+                auto end = theirPos & position_mask;
+
+                // [begin, end) is the writable/readable range for the producer/consumer.

```

Comment:
The only thing I might ask for here and throughout this file is more ASCII-art diagrams. I totally understand the [begin, end) notation, but I feel like this is one of those sufficiently complex solutions where some good ol' ASCII art diagrams with what the circular buffer looks like and arrows into the positions or valid ranges might do a bit of good. You can veto this if you want. I just like ASCII-art opportunities.